src/preact/base-element.js

```diff
@@ -131,6 +141,45 @@ const MATCH_ANY = () => true;
 
 const childIdGenerator = sequentialIdGenerator();
 
+const ONE_OF_ERROR_MESSAGE =
+  'Only one of "attr", "attrs", "attrPrefix", "passthrough", ' +
+  '"passthroughNonEmpty", or "selector" must be given';
+
+/**
+ * @param {!Object<string, !AmpElementPropDef>} propDefs
+ * @param {function(!AmpElementPropDef):boolean} cb
+ * @return {boolean}
+ */
+function checkPropsFor(propDefs, cb) {
+  return Object.values(propDefs).some(cb);
+}
+
+/**
+ * @param {!AmpElementPropDef} def
+ * @return {boolean}
+ */
+const HAS_MEDIA = (def) => def.media;
+
+/**
+ * @param {!AmpElementPropDef} def
+ * @return {boolean}
+ */
+const HAS_SELECTOR = (def) => typeof def === 'string' || def.selector;
+
+/**
+ * @param {!AmpElementPropDef} def
+ * @return {boolean}
+ */
+const HAS_PASSTHROUGH = (def) => def.passthrough || def.passthroughNonEmpty;
+
+/**
+ * Ignore whitespace text nodes.
+ * @param {Node} node
+ * @return {boolean}
+ */
+const IS_REAL_CHILD = (node) =>
```

Comment:
Nit: Can we rename this to `IS_NONEMPTY_TEXT`?

Reply:
good idea, done